File: packages/python/python-shared/src/python_shared/http/resilience/fleet_retry_budget.py
from typing import Dict, Any

class FleetRetryBudget:
    def __init__(self, max_retry_ratio: float = 0.2, min_requests_threshold: int = 10):
        self.max_retry_ratio = max_retry_ratio
        self.min_requests_threshold = min_requests_threshold
        self.total_requests = 0
        self.total_retries = 0

    def record_request(self) -> None:
        self.total_requests += 1

    def record_retry(self) -> None:
        self.total_retries += 1

    def can_retry(self) -> bool:
        if self.total_requests < self.min_requests_threshold:
            return True
        ratio = self.total_retries / self.total_requests
        return ratio <= self.max_retry_ratio

    def get_stats(self) -> Dict[str, Any]:
        ratio = self.total_retries / self.total_requests if self.total_requests > 0 else 0.0
        return {
            "totalRequests": self.total_requests,
            "totalRetries": self.total_retries,
            "ratio": ratio,
        }

File: packages/llm-obs-node-packages/shared-infra/src/http/resilience/fleet-retry-budget.ts
/**
 * @file fleet-retry-budget.ts
 * @description Fleet-Wide Retry Storm Prevention Budgeting Engine.
 * 
 * ALGORITHM & SPECIFICATION:
 * 1. Retry Storm Prevention:
 *    - Tracks total requests (`totalRequests`) and total retry attempts (`totalRetries`).
 *    - Evaluates retry ratio: `ratio = totalRetries / totalRequests`.
 * 2. Dynamic Retry Suppression:
 *    - `canRetry()`: If `totalRequests >= minRequestsThreshold` AND `ratio > maxRetryRatio` (default: 20%),
 *      suppresses retries globally by returning `false`.
 *    - Prevents microservices from compounding downstream outages via retry storms.
 */

export class FleetRetryBudget {
  private totalRequests = 0;
  private totalRetries = 0;

  constructor(
    private readonly maxRetryRatio = 0.2,
    private readonly minRequestsThreshold = 10
  ) {}

  public recordRequest(): void {
    this.totalRequests++;
  }

  public recordRetry(): void {
    this.totalRetries++;
  }

  public canRetry(): boolean {
    if (this.totalRequests < this.minRequestsThreshold) {
      return true;
    }
    const ratio = this.totalRetries / this.totalRequests;
    return ratio <= this.maxRetryRatio;
  }

  public getStats(): { totalRequests: number; totalRetries: number; ratio: number } {
    const ratio = this.totalRequests > 0 ? this.totalRetries / this.totalRequests : 0;
    return {
      totalRequests: this.totalRequests,
      totalRetries: this.totalRetries,
      ratio,
    };
  }
}
